text/3742-mutate-non-mutable-locals.md

```diff
@@ -0,0 +1,249 @@
+- Feature Name: `mut_non_mut`
+- Start Date: 2024-12-13
+- RFC PR: [rust-lang/rfcs#3742](https://github.com/rust-lang/rfcs/pull/3742)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+the `mut` keyword has two meanings currently:
+ - When applied to a reference, it means the reference is not aliased
+ - When applied to a local binding, it means that the binding is allowed be re-assigned or borrowed using `&mut`
+
+The first of these is fundamental to Rust's safety guarantees and enables some optimizations.  The second,
+however, is more of a lint: the error message may help catch mistakes, but ignoring it doesn't violate
+type safety or subvert any of Rust's global correctness guarantees.
+
+This RFC proposes that the re-assignment or creation of a `&mut` to a local variable not
+marked with `mut` be changed from a hard error to a deny-by-default lint diagnostic (called
+`mut_non_mut` for the sake of discussion).
+
+# Motivation
+[motivation]: #motivation
+
+This idea has been discussed previously:
+
+ - [2014 Blog post](https://smallcultfollowing.com/babysteps/blog/2014/05/13/focusing-on-ownership/)
+ - [2021 IRLO thread](https://internals.rust-lang.org/t/lack-of-mut-in-bindings-as-a-deny-by-default-lint/15818)
+ - [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/Idea.3A.20Downgrade.20.22immutable.20variable.E2.80.9C.20errors.20to.20lints)
+
+One benefit of this change is reduced friction: adding and removing of the need for `mut` on a given local
+can happen frequently when refactoring, similar to how variables and functions may change from used to unused
+and back.  In this context, a fatal error for `mut_non_mut` can be a flow interruption.  Yes, adding
+`mut` in all the right places is ideally something you should do _eventually_, just as you should
+delete any unused variables/functions, but it's not something that needs to prevent you from running
+the program.
+
+Another benefit is towards language consistency/explainability: having `mut_non_mut` be a configurable lint
+makes it reasonable that it will sometimes have false positives/negatives.  On the other hand, making it a
+hard error adds its rules to the language sematics (and worse: conflates them with borrow checking):
+ - Does `mut` mean mutable?  Surely not, since there exists interior mutability.
+ - But `mut` doesn't just mean "exclusive," either: local variables are _already_ exclusive
+   by nature, so wouldn't need `mut` under that definition.
+ - If you can't get a `&mut` to a local declared without `mut`, how does `drop` do it?
+ - There is also the "move loophole," where moving out of a binding lets you mutate it:
+   ```rs
+   let not_mutable = vec![1, 2, 3];
+   let mut not_mutable = not_mutable;
+   not_mutable.push(4);
+   ```
+ - But that loophole _doesn't_ work for `move` closures, which ostensibly do the same thing:
```

Comment:
I'm personally not convinced by this argument: The snipped below doesn't include the `mut` keyword anywhere, so why would it enable mutation? (And it compiles just fine if you add `let mut not_mutable = not_mutable;` to the closure body).

Reply:
I got a bit too fancy with the IIFE.  If you assign to a temporary you can see the `mut`:

```rs
let not_mutable = vec![1, 2, 3];
let mut f = move || { // `f` is `mut` and owns `not_mutable` after the move
    not_mutable.push(4); // still rejected by the compiler
};
f();
```